### ISSUE TITLE
BUG - sev-3 - All - With RQ (and even more so after HSP) we will have both frontend and backend caches, and they aren't the same

### ISSUE THREAD

**Issue by TKDickson** [department-of-veterans-affairs/va-mobile-app#8584]
## What happened?

On the backend, we cache these responses (if they're not errors):
Messages for 20 minutes
Appointments, and claims/appeals for 30 minutes

With react query, we add a cache on the frontend of 5 seconds to all API calls that have been migrated.

With home screen personalization, we change the frontend cache to 5 minutes for a bunch of stuff on the home screen, but importantly - including all of the stuff that's cached for longer on the backend.

## Specs:

- Device:
- OS:
- User Account:
- Accessibility State:

## Desired behavior

We should have a meeting and make some decisions here, but what I think we should do is:

- drop react query retries to zero. _(arguably, not related to caching)_
- Drop BE caching all together and cache on frontend (to avoid loading screens for no reason)
- Contemplate a realistic cache time and identify cache break points and document them.
- The only thing to add to that is deciding what to do about pagination problems, or if we even care about those

## Bug Severity - BE SURE TO ADD THE SEVERITY LABEL
 
See [Bug Tracking](https://department-of-veterans-affairs.github.io/va-mobile-app/docs/QA#issue-severity) for details on severity levels

- Impact: High Low 
- Frequency: High Low 

- 3 - Low
- 2 - High
- 1 - Critical

## Screen shot(s) and additional information

<details>
  <summary>Full JSON response for services related to issue (expand/collapse)</summary>
  

</details>

## Ticket Checklist
- [ ] Steps to reproduce are defined
- [ ] Desired behavior is added
- [ ] Labels added (front-end, back-end, global, Health, relevant feature, accessibility, etc)
- [ ] Estimate of 1 added (for front-end tickets)
- [ ] Added either to the relevant feature epic (if found during new feature implementation), or the relevant team's bug epic ([Global](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/8302), [Health & Benefits](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/8303), [API](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/8304), [QART](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/8305))

**Comment by TKDickson** [comment]
Had a meeting in which FE and BE generally agreed:
* Caching should move to the FE, and initial caches should be set relatively low (ex: 5 minutes)
* Moving caching to the FE can happen in FE-only code, for now
* Still TBD on "what all scope exactly is happening when" <-- this will require input from @ajsarkar28 when he's back in, for how it intersects with HSP
* Eventually, BE would want to version things (to have versions without the BE caching, for ease of things going forward), they're going to write tickets/epic/etc for their work for that

@jperk51 @dumathane feel free to add what I missed.

**Comment by theodur** [comment]
I moved the work for deciding query cache times to a new ticket and assign Binny to it
https://github.com/department-of-veterans-affairs/va-mobile-app/issues/8638

**Comment by theodur** [comment]
The work for this is complete in a [draft PR](https://github.com/department-of-veterans-affairs/va-mobile-app/pull/8635). I'm just waiting on what the BE team says in their conversation with Binny, which could lead to some minor changes in the PR. There's also one failing unit test that I need to troubleshoot. I'll fix that tomorrow

**Comment by theodur** [comment]
The PR for this has a lot more changes than expected in order to account for pagination issues. The most notable issue I saw was the duration of the API calls where we're manually breaking the backend cache. The time is significantly longer when we break the cache without any additional changes. Here's an example when paginating through appointments:

[BustedCacheAppointmentsFetch.webm](https://github.com/department-of-veterans-affairs/va-mobile-app/assets/9042882/99652602-fdb7-4291-a3b3-03c3855b4c2b)

The cache busted endpoints needed to be updated to fetch all the data upfront, similar to prescriptions. The query for prescriptions fetches up to 5000 entries upfront, which is what I set the count to for the other endpoints also (appointments, secure messages, claims, vaccines). As a result, I had to update how we handle pagination on the FE.

**Comment by theodur** [comment]
I moved the RQ retry change into another ticket/PR after the discussion we had in the React Query retro
https://github.com/department-of-veterans-affairs/va-mobile-app/issues/8657

**Comment by TKDickson** [comment]
Relevant changes:

> Busting the backend cache for appointments, claims & appeals, secure messages, and vaccines to ensure we're always retrieving fresh data whenever we fetch from those APIs
> Managing pagination in the app for the above APIs
> During my testing, I noticed for the vaccines and claims lists, the screens weren't auto-scrolling to the top when cycling through pages in the pagination, so I added that functionality for those screens.
> Pre-loading past appointments and past claims (to assist with the speed of those features loading)

So the behavior with this, the normal HSP changes, and retries to 0 (which is in develop):

- Claims/appeals, appointments, and messages should cache on the frontend for 5 minutes (or until a mobile client cache-busting action, like cancelling an appt on the mobile client). Vaccines will cache, presumably, for 5 seconds (the default for RQ).
- Changes outside of the mobile app (scheduling an appt on the website; physician sends a new secure message) will be reflected in the next API call (either the length of the cache, or through a re-trying action, like leaving and re-entering the activity)
- Pagination (& the corresponding auto-scrolling) should all still work

I have *no idea* what will happen if you're in the middle of pagination when the cache needs to update (ex: scrolling through upcoming appointments and a new appointment is scheduled). I am going to attempt to do this test case for all of the impacted activities, although I'll need to mock it for vaccines for certain.

Claims and appeals
- [ ] Cache time without changes
- [ ] Take a cache-busting action -- not possible in mobile app for claims
- [ ] Cache time check with changes (ex: adding a new claim)
- [ ] Pagination and auto-scrolling
- [ ] Changes while in the midst of pagination, at a cache refresh time

Messaging
- [ ] Cache time without changes
- [ ] Take a cache-busting action (send or read a message)
- [ ] Cache time check with changes (ex: sending a new message)
- [ ] Pagination and auto-scrolling
- [ ] Changes while in the midst of pagination, at a cache refresh time

Appointments
- [ ] Cache time without changes
- [ ] Take a cache-busting action (cancel an appointment)
- [ ] Cache time check with changes (ex: scheduling a new appt)
- [ ] Pagination and auto-scrolling
- [ ] Changes while in the midst of pagination, at a cache refresh time

Vaccines
- [ ] Cache time without changes
- [ ] Take a cache-busting action -- not possible in mobile app
- [ ] Cache time check with changes (do not have the ability to do real prep, will need to mock with Charles)
- [ ] Pagination and auto-scrolling
- [ ] Changes while in the midst of pagination, at a cache refresh time

**Comment by TKDickson** [comment]
Pagination & caching/changes isn't a concern with vaccines (we load the full list of vaccines, at a 5000 vaccines 'limit', then page through it). 

I've found a number of not-caused-by-this-dev issues, including #8706, #8689, and reopening #4762. Investigating and writing those up is slowing testing for this overall (and also blocking it, in some cases).

**Comment by TKDickson** [comment]
- [Reading a message isn't updating the count on the health page or home screen](https://drive.google.com/file/d/1uiILAt8eKSM0NDhFQ1fRUwbtT2IXQWUk/view?usp=sharing). (Android, +226). I have to wait until the cache time expires, then the count updates after the full API rigamarole again. 
- I think we could probably limit the "pre-loading past appointments and past claims" to JUST when the user first logs in. (not re-calling when cache-busting action is taken, nor re-calling when the FE cache expires). 
-

**Comment by TKDickson** [comment]
Oh and I found a few more not-this-dev issues: #8724, #8725, and #8726.

**Comment by theodur** [comment]
The first issue of the unread message count not updating was caused by #8577; that ticket was completed last week. Those changes are in my branch now after merging `develop` into the feature XL branch, and I confirmed the count is being updated now. The second issue should be fixed now also. There may be a way to fix the 3rd issue with the 502 response, I'd just need to know what the error object is that's being returned in that scenario

**Comment by theodur** [comment]
After tinkering around with the claims API error issue, I think we should go ahead and make a separate ticket for that. We'll need to find a good way to go about this in the FE side. The first 2 issues are fixed though, so sending this back to @TKDickson

**Comment by TKDickson** [comment]
Confirmed that I'm not seeing the past appt and past claims APIs being recalled when re-navigating to the home screen after 5 minutes as judy.morrison.

We're out of unread messages for our staging user, and I'm having some issues with my computer recognizing my PIV card reader today, so will try a couple more things before needing to reach out to someone else to assist with data prep to be able to test the other fix.

Planning on writing up the third thing separately.

**Comment by TKDickson** [comment]
#8752 for the third thing. Confirmed that reading messages is reflecting right away on home screen as well. This is ready to merge.

**Comment by TKDickson** [comment]
Merged, closing.
